### PR TITLE
Cleanup vehicle drawing code part 4

### PR DESCRIPTION
**Pull request by spacek531**

### CODE REVIEW

File: src/openrct2/object/RideObject.h [hunk 1]
```diff
@@ -50,8 +50,8 @@ class RideObject final : public Object
     void ReadLegacyVehicle(IReadObjectContext* context, OpenRCT2::IStream* stream, rct_ride_entry_vehicle* vehicle);
 
     void ReadJsonVehicleInfo(IReadObjectContext* context, json_t& properties);
-    std::vector<rct_ride_entry_vehicle> ReadJsonCars(json_t& jCars);
-    rct_ride_entry_vehicle ReadJsonCar(json_t& jCar);
+    std::vector<rct_ride_entry_vehicle> ReadJsonCars([[maybe_unused]] IReadObjectContext* context, json_t& jCars);
+    rct_ride_entry_vehicle ReadJsonCar([[maybe_unused]] IReadObjectContext* context, json_t& jCar);
```
Comment (duncanspumpkin) [inline src/openrct2/object/RideObject.h:54]: i assume this is for something in the future?

File: src/openrct2/ride/water/SplashBoats.cpp [hunk 2]
```diff
@@ -1276,7 +1276,7 @@ void vehicle_visual_splash_boats_or_water_coaster(
     }
 
     session.CurrentlyDrawnEntity = vehicleToPaint;
-    imageDirection = ((session.CurrentRotation * 8) + vehicleToPaint->sprite_direction) & 0x1F;
+    imageDirection = OpenRCT2::Entity::Yaw::MaskYaw(OpenRCT2::Entity::Yaw::YawFrom4(session.CurrentRotation) + vehicleToPaint->sprite_direction);
```
Comment (duncanspumpkin) [inline src/openrct2/ride/water/SplashBoats.cpp:1279]: This doesn't look too nice perhaps an `Yaw::Add` would be nicer:
```
imageDirection = OpenRCT2::Entity::Yaw::Add(vehicleToPaint->sprite_direction, OpenRCT2::Entity::Yaw::YawFrom4(session.CurrentRotation));
```
I feel that is easier to understand than talking about masks.

File: src/openrct2/entity/Yaw.hpp [hunk 3]
```diff
@@ -18,28 +18,46 @@ namespace OpenRCT2::Entity::Yaw
     // smallest precision is 4 frames
     constexpr const uint8_t PrecisionOffset[] = { 3, 2, 1, 0, 0, 0, 0, 0 };
 
-    constexpr int32_t YawTo4(int32_t yaw)
+    constexpr [[nodiscard]] int32_t MaskYaw(int32_t yaw)
```
Comment (duncanspumpkin) [inline src/openrct2/entity/Yaw.hpp:21]: See other comments i think an AddYaw would be better.

File: src/openrct2/object/RideObject.h (hunk 1 above)
Comment (spacek531) [reply src/openrct2/object/RideObject.h:54]: Yes, the feature used to be in this PR but I changed how it was in a later PR so I just removed it from this one.

File: src/openrct2/ride/VehiclePaint.cpp [hunk 4]
```diff
@@ -3136,9 +3052,9 @@ static constexpr const vehicle_sprite_func PaintFunctionsByPitch[] = {
     VehiclePitchDiagDown12,
     VehiclePitchDiagDown25,
     VehiclePitchDiagDown60,
-    VehiclePitchInvertingDown60,
```
Comment (duncanspumpkin) [inline src/openrct2/ride/VehiclePaint.cpp:3139]: I take it this was just a bunch of mistaken names?

File: src/openrct2/entity/Yaw.hpp [hunk 5]
```diff
@@ -18,28 +18,46 @@ namespace OpenRCT2::Entity::Yaw
...
+    constexpr [[nodiscard]] int32_t YawToPrecision(int32_t yaw, uint8_t precision)
     {
         return yaw >> PrecisionOffset[precision];
     }
-    constexpr uint8_t NumSpritesPrecision(uint8_t precision)
+    constexpr [[nodiscard]] uint8_t NumSpritesPrecision(uint8_t precision)
     {
         return 4 << precision;
     }
+    constexpr uint8_t [[nodiscard]] PrecisionFromNumSprites(uint8_t numSprites)
```
Comment (duncanspumpkin) [inline src/openrct2/entity/Yaw.hpp:53]: What is the planned use for this? Do we really need something like this perhaps a table would be easier for most readers to understand.

File: src/openrct2/ride/VehiclePaint.cpp (hunk 4 above)
Comment (spacek531) [reply src/openrct2/ride/VehiclePaint.cpp:3139]: yes

File: src/openrct2/ride/VehiclePaint.cpp [hunk 6]
```diff
@@ -2401,7 +2325,8 @@ static void VehiclePitchUp150(
     if (vehicleEntry->sprite_flags & VEHICLE_SPRITE_FLAG_VERTICAL_SLOPES)
     {
         int32_t boundingBoxNum = (YawTo4(imageDirection)) + 88;
-        int32_t spriteNum = (((YawTo4(imageDirection)) + 96) * vehicleEntry->base_num_frames)
+        int32_t spriteNum = (((YawTo4(imageDirection)) + 32 + vehicleEntry->NumRotationFrames() * 2)
```
Comment (duncanspumpkin) [inline src/openrct2/ride/VehiclePaint.cpp:2328]: are you sure it shouldn't be NumRotationFrames()*3 and drop the +32 or am i misunderstanding how this works.

Comment (spacek531) [reply src/openrct2/ride/VehiclePaint.cpp:2328]: The sprites that are not vertical are fixed at 4 frames per rotation

**Review by duncanspumpkin — APPROVED** [review]
This looks like a really great refactoring. Its starting to make much more sense.

**Comment by duncanspumpkin** [comment]
Looks like the non msvc compilers don't handle [[nodiscard]] correctly. You'll need to move it to the left of `constexpr`

**Pull request merged**